Implement LRA Theory Solver to be used to handle relational assumptions in new assumptions

TiloRC opened the pull request:
Continued in https://github.com/sympy/sympy/pull/25544

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

Adds a linear real arithmetic theory solver to logic/algorithms. It has the ability to check if conjunctions of equalities and strict and non-strict inequalities are satisfiable and is designed to work with encoded cnf objects.  

#25544 integrates the code from this pull request with dpll2.py.

I haven't done any tests, but I suspect the speed bottle neck is how it processes inequalities in `from_encoded_cnf`. Ideally the inequalities would come preprocessed. 

#### Statistics 

To better understand what was slowing down, I measured how long each function in the LRASolver was taking to run in random tests. `from_encoded_cnf()` does preprocessing once for each problem, `assert_enc_boundry()` activates each constraint in the problem, and `check()` is where the main algorithm runs to check if the problem is satisfiable. 

EDIT: Since running the tests, I have made improvements to `from_encoded_cnf()` which speed it up dramatically. 

##### Problems of Varying Sizes

|            | `from_encoded_cnf()` | `assert_enc_boundry()` | `check()` | Total Time (s) | % sat |
|---------------------|--------------------|----------------------|-------|-------|-------|
| 2 var, 4 const  | 12.4               | 0.129                | 0.922 | 13.5  | 51    |
| 4 var, 8 const  | 43.0               | 0.265                | 7.00  | 50.2  | 41    |
| 8 var, 16 const  | 150                | 0.555                | 64.7  | 214.77| 28    |
| 16 var, 32 const | 584 | 1.28 | 1000 | 1590 | 14

##### Integer vs Rational Coefficients

|            | `from_encoded_cnf()` | `assert_enc_boundry()` | `check()` | Total Time (s) | % sat |
|---------------------|--------------------|----------------------|-------|-------|-------|
| integer coeff | 43.0               | 0.265                | 7.00  | 50.2  | 41    |
| rational coeff  | 60.6                | 0.264                | 8.80  | 69.6 | 41    |

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

NO ENTRY

<!-- END RELEASE NOTES -->

TiloRC committed:
Move find_feasible code from #25277 to new PR

TiloRC committed:
semi-working implementation of the paper

Still needs to be tested more thoughroughly.
There are probably more bugs. Currently in the
processes of getting rid of bug.

TiloRC committed:
small change

Upabjojr committed:
added debug printer to LRASolver.check

TiloRC committed:
fix -A[i, j]

TiloRC committed:
rename: lra.beta --> lra.assign

TiloRC committed:
check: rename local A to M to avoid confusion with self.A

TiloRC committed:
check: make basic, nonbasic variables

basic and nonbasic are distinct from self.basic and self.nonbasic

TiloRC committed:
implement _debug_internal_state_printer

TiloRC committed:
Changed how pivoting worked

Previously the code was doing pivoting as described
in Ferguson's notes. I moved to a more standard way
of pivoting.

TiloRC committed:
Fixed bug!!!!!!

TiloRC committed:
Clean up some tests

TiloRC committed:
Clean up code

Deleted commented out code.

TiloRC committed:
reorganize variables

self.basic and self.nonbasic variables have been renamed
to self.slack and self.nonslack.

TiloRC committed:
Deleted unused line: j = nonbasic[xj]

TiloRC committed:
check: reorganize basic and nonbasic variables

TiloRC committed:
return variable assignment when SAT

TiloRC committed:
Standardize return type of assert_con to always be tupple

TiloRC committed:
find_feasible: include conflict clause in retured tuple

TiloRC committed:
LRASolver: Add variables parameter to preproccesing

TiloRC committed:
LRASolver.__init__: add assert statement to check A formatting

A should have a negative identity matrix at the end for slack
variables.

TiloRC committed:
remove obsolete code

TiloRC committed:
LRASolver.preprocess: Use dummy variables

TiloRC committed:
LRASolver.preprocess: handle inequalities not in standard form

TiloRC committed:
Fix random testing by setting sparcity to 0

TiloRC committed:
print assignments in debugger

TiloRC committed:
add outline for new method: from_encoded_cnf

smichr commented:
Do you have (or have you seen) a nice example of where the solution from a system of strict inequalities is significantly different from one where the inequalities are not? And by significantly different I mean not including the case where `a<b` is false but `a<=b` is true.

TiloRC committed:
improve rand problem generation: make simpler problems

The coefficients of variables in constraints are ints by default
instead of Rationals now.

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -486,3 +490,127 @@ def test__pt():
+# lra.assert_con(s1 <= 1)
+# lra.assert_con(s2 >= -3)
+# lra.check()
+
+
+
+
```
```suggestion
```

smichr committed:
pep8

TiloRC committed:
added assert within check():

assignments for x must always satisfy Ax = 0

TiloRC committed:
fix bug with pivot

basic variables should be negative

TiloRC committed:
fix pivot

I think this should work. I don't think I'll change it again.

TiloRC committed:
thousands of random tests all pass

TiloRC committed:
Merge branch 'LRA-solver' of https://github.com/TiloRC/sympy into LRA-solver

TiloRC commented:
> Do you have (or have you seen) a nice example of where the solution from a system of strict inequalities is significantly different from one where the inequalities are not?

I'm not exactly sure what you mean by this. Does this count as significantly different?

Has no solutions:

x+y>0,  x<0,  y<0

Has a solution at x=0, y=0:

x+y>=0,  x=<0,  y=<0

Why do you ask? Is it because you want to try to implement the find feasible function using the code from the linear programing PR? I know I said that wasn't possible, but it might be. It's just that the approach the paper I'm using to implement this PR won't work.

Basically what this paper does is that strict inequalities like x < 0 become x + delta <= 0 where delta is an infinitely small quantity. The problem with doing this in the in the linear programing PR is that I don't think multiplication/division between terms like  x + delta is well defined. 

If you're interested in solutions for linear programing problems with strict inequalities, I don't think this is possible. For example, it's not possible to get a maximum value for x subject to x < 0.

oscarbenjamin commented:
> Has no solutions:
> 
> x+y>0, x<0, y<0
> 
> Has a solution at x=0, y=0:
> 
> x+y>=0, x=<0, y=<0

One way to approach this is to rewrite strict inequalities like:
```
x > 0    -->   x >= 0, x != 0
x + y > 1   -->   z = x + y - 1, z >= 0, z != 0
```
The unequalities would not be included in the linear programming problem but they could be checked afterwards: if the only feasible value of z was zero then that can be rejected. Otherwise the condition `z != 0` can give us a symbolic condition on the variables (`x + y - 1 != 0`) that can be used to constrain any conclusions generated using `z >= 0`.

TiloRC commented:
It seems like there are probably a lot of ways to go about dealing with strict inequalities. I think I'm probably going to go with what the paper I'm implementing does. This paper is used by Z3 so the approach must be pretty fast.

smichr commented:
> Why do you ask?

It just seems to me that if you have a solution for non-strict inequalities, if you change them all to strict inequalities the solution will almost be the same -- but you will be located eps away from the solution. So if an inequality was y < 3 and the solution contained y = 3 then you would have to make the solution 3 - var('eps', positive=True). But as someone says [here](https://math.stackexchange.com/questions/2673917/can-a-linear-program-have-strict-inequalities) a minimization not on the boundary is strange -- you have to tell what values *can* be part of the solution (hence closed space).

Perhaps what is different is that in checking feasibility, one wants to know if you *can* have a solution somewhere, not necessarily max or min. But that seems similar to the max/min problem since if you can have a max or min, the system is feasible. It will be interesting to compare the approaches when you are done.

TiloRC commented:
> Perhaps what is different is that in checking feasibility, one wants to know if you can have a solution somewhere, not necessarily max or min.

This is my understanding. 

> It will be interesting to compare the approaches when you are done.

Cool! I presume you will implement something similar then?

TiloRC commented:
I'm not sure where the best place to put the code in this PR is. Maybe a better home for it would be sympy.logic.inference.

smichr commented:
> presume you will implement something similar then?

no -- just the feasibility step of `lp` (or whatever it will be called). I am thinking of sticking with the scipy convention and making an optimize module and call the function `linprog`. I'm not sure if that will be confusing, however, since our inputs are different than theirs.

smichr commented:
> Maybe a better home for it would be sympy.logic.inference

hmm...Instead of an optimize maybe a module for inequality_systems could be added. Then `lp` and `is_feasible` could be there.

TiloRC commented:
> Then `lp` and `is_feasible` could be there.

It might make sense to get rid of `is_feasible` because [`satisfiable`](https://docs.sympy.org/latest/modules/logic.html#sympy.logic.inference.satisfiable) will be able to do everything `is_feasible` can do and more.

smichr commented:
> It might make sense to get rid of `is_feasible`

I meant whatever function you are working on.

TiloRC commented:
> no -- just the feasibility step of lp (or whatever it will be called).

I think it would make most sense to raise an exception if the input contains a strict inequality in the context of linear programing. That's what other software does.

TiloRC commented:
> I meant whatever function you are working on.

`satisfiable` is an existing function. My idea is rather than making a new function just make `satisfiable` handle inequalities.

TiloRC commented on sympy/solvers/inequalities.py:
```diff
@@ -982,3 +988,338 @@ def reduce_inequalities(inequalities, symbols=[]):
+
+            # assignments for x must always satisfy Ax = 0
+            X = Matrix([self.assign[v] for v in self.col_index])
+            assert all(abs(val) < 10**(-10) for val in M*X) # TODO: Maybe return None if this cond can't be met?
```
Maybe we can have a variable that's set when the tests are being run so that these assert statements are only run during the tests. Maybe pytest has a way to this.

asmeurer replied:
@oscarbenjamin any thoughts on how this could be done? Is there a variable that gets set when the tests are run?

oscarbenjamin replied:
Why not have a parameter like `check=False` and then call the function with check=True during the test code? Then it is easier to enable checking in any situation rather than just the test suite.

The kind of thing that I would potentially want an assert like this for is to pick up on problems that might arise in cases that are *not* tested e.g. because user code does something unanticipated. If the asserts only run during testing then they can just be part of the test suite.

TiloRC commented on sympy/solvers/inequalities.py:
```diff
@@ -982,3 +988,338 @@ def reduce_inequalities(inequalities, symbols=[]):
+             or self.assign[b] > self.upper[b]]
+
+            if len(cand) == 0:
+                return "SAT", self.assign
```
Use True/False for "SAT"/"UNSAT"

TiloRC committed:
progress so far

TiloRC committed:
strict inequalities now work

However, conflict clauses and assignments no longer work.

TiloRC committed:
Test cases seem to be passing

TiloRC committed:
bug fix: _assert_upper/lower never returns sat

This was causing a bug. I don't know why the pseudo code
from paper was doing it that way.

TiloRC committed:
conflict clauses are being generated again

This is mostly to check if everything is working.
I have yet to implement them in a way that works
nicely with SAT solver.

TiloRC commented:
This comment is mostly for my own future reference.
 
I'm not totally sure what to do with equality. Here are some options: 

### in a preprocessing step split each equality into two inequalities

I can't do this because equality is not necessarily equivalent to two inequalities. Inequalities imply their arguments are real numbers whereas equalities do not. 

### Add clauses that convey that Q.eq(x, y) implies x <= y and x>=y

This would slow things down. Also possibly a bit involved.  

### Have LRASolver deal with equality internally 

The difficulty here is that generating conflict clauses becomes very difficult as conflict clauses are made up of only inequalities. Perhaps I can keep track of whether the current boundaries for each variable are the result of an equality or not. 

I think the LRASolver should probably deal with equality internally. This will be a bit messy, but I think it's the best option.

TiloRC commented:
Things I want to get done by the meeting tomorrow: 
- equality should be handled 
- conflict clauses should be literals
- clean up all the junk/old code that's no longer needed 
- if I have time, start thinking about back tracking

TiloRC committed:
conflict clauses now returned as encoded literals

TiloRC committed:
created docstring for _pivot

TiloRC committed:
conflict clauses are now negative

TiloRC committed:
LRASolver: added docstring and comments to from_encoded_cnf

TiloRC committed:
delete print statements

TiloRC committed:
depreciate functions made obsolete by from_encoded_cnf

TiloRC committed:
write something for check and assert_enc_boundry docstrings

TiloRC committed:
skip doctests because of nondetermism

TiloRC committed:
update and remove tests for depeciated code

TiloRC committed:
move test_random_problem to bottom of testing file

TiloRC committed:
remove old print statement

TiloRC committed:
add test for pivot operation

TiloRC committed:
remove changes from solvers module

all code should be in the logic module now

TiloRC committed:
clean up any changes I missed in Solver module

TiloRC committed:
remove any changes I missed in solvers module

fogot to add inequalities file; only added testing file in last
commit.

TiloRC committed:
remove unused imports and fix spacing

TiloRC committed:
remove costly rref assertion in _pivot

Tests have been added for this in
test_lra_theory.py.

TiloRC committed:
spacing

TiloRC committed:
add returns section to check docstring

TiloRC committed:
Split long lines over multiple lines

TiloRC committed:
check uses True/False instead of SAT and UNSAT

TiloRC committed:
add returns section to assert_enc_boundry docstring

TiloRC commented on sympy/logic/tests/test_lra_theory.py:
```diff
@@ -0,0 +1,306 @@
+    # A = -A # the identity matrix should be negative
+    # lra = LRASolver(A, [s1, s2], [x, y])
+    # assert lra.check() == ('SAT', {x: 0, y: 0, s1: 0, s2: 0})
+    # assert {v: lra.assign[v] for v in lra.all_var} == {x:0, y:0, s1:0, s2:0}
+    #
+    # assert lra.assert_con(x <= -4) == ("OK", None)
+    # assert lra.check() == ('SAT', {x: -4, y: 0, s1: 4, s2: -4})
+    # assert {v: lra.assign[v] for v in lra.all_var} == {x:-4, y:0, s1: 4, s2:-4}
+    #
+    # assert lra.assert_con(x >= -8) == ("OK", None)
+    # assert lra.check() == ('SAT', {x: -4, y: 0, s1: 4, s2: -4})
+    # assert {v: lra.assign[v] for v in lra.all_var} == {x:-4, y:0, s1:4, s2:-4}
+    #
+    # # note that this is the first time a pivot is used
+    # assert lra.assert_con(s1 <= 1) == ("OK", None)
+    # assert lra.check() == ('SAT', {x: -4, y: -3, s1: 1, s2: -7})
+    # assert {v: lra.assign[v] for v in lra.all_var} == {x:-4, y:-3, s1:1, s2:-7}
+    #
+    # assert lra.assert_con(s2 >= -3) == ("OK", None)
+    # assert lra.check() == ('UNSAT', {s1 <= 1, x <= -4, s2 >= -3})
+    #
+    #
+    # r1 = x <= 0
+    # r2 = y <= 0
+    # r3 = s1 >= 2
+    # equations = [Eq(s1, x+y)]
+    # A, _ = linear_eq_to_matrix(equations, [x, y, s1])
+    # A = -A  # the identity matrix should be negative
+    # lra = LRASolver(A, [s1], [x, y])
+    # lra.assert_con(x <= 0)
+    # lra.assert_con(y <= 0)
+    # lra.assert_con(s1 >= 2)
+    # assert lra.check() == ('UNSAT', {s1 >= 2, x <= 0, y <= 0})
+    #
+    #
+    # # test potential edge case
+    # r1 = x <= 1
+    # r2 = -x <= -5
+    # lra = LRASolver(Matrix(), [], [x])
+    # lra.assert_con(x <= 1)
```
This code uses depreciated functions. I'm not sure if I should convert it into code that uses the new functions or just get rid of it. I feel like it's kind of pointless because test_random_problems() should be able to handle all of the cases here.

TiloRC committed:
clean up print statements

TiloRC committed:
remove newline at file end

TiloRC committed:
attempt to remove unicode

TiloRC committed:
remove unnesesary set comprehension with list comp

TiloRC committed:
removed unassigned var: Mi, Mj

TiloRC committed:
remove unassigned var: r = random()

TiloRC committed:
fix Unnecessary generator: rewrite as set comprehension

TiloRC committed:
fix: Unnecessary generator - rewrite as a set comprehension

TiloRC commented on sympy/logic/tests/test_lra_theory.py:
```diff
@@ -0,0 +1,297 @@
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.ge(x, 0))) +1) == ("OK", None)
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.ge(x, 1))) +1) == ("OK", None)
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.le(x, -1))) + 1) == ('UNSAT', {-(enc.symbols.index((Q.le(x, -1))) + 1), -(enc.symbols.index((Q.ge(x, 1))) +1)})
+    # assert lra.assert_con(Q.ge(x, 0)) == ("OK", None)
+    # assert lra.assert_con(Q.ge(x, -1)) == ('SAT', {x: 0, y: 0})
+    # assert lra.assert_con(Q.le(x, -1)) == ('UNSAT', {x <= -1, x >= 0})
+    #
+    # m = Matrix()
+    # lra = LRASolver(m, [], [x, y])
```
This commented out code code uses depreciated functions. I'm not sure if I should convert it into code that uses the new functions or just get rid of it. I feel like it's kind of pointless because test_random_problems() should be able to handle all of the cases here.

TiloRC committed:
reword check docstring

TiloRC committed:
WIP: start implementing backtracking

TiloRC committed:
LRA Solver: add reset_bounds()

TiloRC committed:
ready LRASolver for SAT integration

- return SAT without having to call Check() when possible
- handle assering ints without an encoding
- fix bug with Q.ne

TiloRC commented:
https://github.com/joerick/pyinstrument

asmeurer commented on sympy/logic/algorithms/lra_theory.py:
```diff
@@ -0,0 +1,667 @@
+        return Q.le(var, const)
+
+
+class Boundry:
```
This could just use a dataclass https://docs.python.org/3/library/dataclasses.html

TiloRC committed:
fix bugs introduced by changes from the integration PR

TiloRC committed:
30x speed up in from_enc_cnf

TiloRC committed:
added profiling code to testing

TiloRC commented:
Continued in #25544. Thought it was confusing to have PR with overlapping code.
